### PR TITLE
Existing Mnemonic

### PR DESCRIPTION
**Pull request by CarlBeek**
Adds support for existing mnemonics (with automatic language detection) and deriving keys from a non-0 index to start.

#94 and #122 are others' attempts at this, but former doesn't check the mnemonic and the latter doesn't offer start-indices.

### CODE REVIEW

**Comment by unexpectedbit** [comment]
It generates different keys when inputting the mnemonic phrase (with no password) that was generated using the existing tool from master branch.

**Comment by torfbolt** [comment]
@unexpectedbit did you compare to keys generated with the current master, or with a previous master? Afaik there were recent changes to the key derivation algorithm, so old keys might not be reproducible even without this PR.

**Comment by unexpectedbit** [comment]
I get the following:

_Current master branch:_

Input:
- Validators: 1
- Mnemonic language: english
- Network: mainnet
- Password: 1

 Output:
- Mnemonic: gossip into stairs acid nest escape cross gate dragon chuckle child minute clarify absorb margin town cabbage barrel embrace release banner shift cargo eight
- Pubkey: b9a37c08ab90b4c684ab7dcfc2ff0dde82e37598cf8bde018874bc4098d1c693f9eaf5e50ccb9657806b7f3884cb5f90
- Path: m/12381/3600/0/0/0

_Current existing_mnemonic branch:_

Input:
- Mnemonic: gossip into stairs acid nest escape cross gate dragon chuckle child minute clarify absorb margin town cabbage barrel embrace release banner shift cargo eight
- Mnemonic password:
- Index:
- Validators: 1
- Network: mainnet
- Password: 1

Output:
- Pubkey: a40747db76e01457348b94c8c4bc242fa22a292204aefab60e66b112a345d8f49e686b7bb39095da9de954b92ffa56b6
- Path: m/12381/3600/0/0/0

**Comment by unexpectedbit** [comment]
After manually modifying the master branch files with existing_mnemonic branch changes, the correct (same) keys are generated!

**Comment by rev3ks** [comment]
This may be due to the changes to the key derivation algorithm mentioned above, but I was unable to create the same keys, I was however able to do so using the fork that @torfbolt created.

**Comment by unexpectedbit** [comment]
@rev3ks Did you try generating the keys using the "_existing_mnemonic_" branch files or did you modify "_master_" branch files with code changes from this PR? I think that the rest of code in files of "_existing_mnemonic_" branch is obsolete and therefore generates different keys.

**Comment by rev3ks** [comment]
I used this branch not master and ran `./deposit.sh existing-mnemonic --mnemonic "..." --num_validators 1 --chain medalla`.

File: tests/test_cli.py [hunk 1]
```diff
@@ -41,9 +41,12 @@ def get_mnemonic(language: str, words_path: str, entropy: Optional[bytes]=None)
         os.mkdir(my_folder_path)
 
     runner = CliRunner()
-    inputs = ['5', 'english', 'mainnet', 'MyPassword', 'MyPassword', 'fakephrase']
+    inputs = [
+        'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about',
+        '2', '2', '5', 'mainnet', 'MyPassword', 'MyPassword', 'yes', 'yes']
     data = '\n'.join(inputs)
-    result = runner.invoke(main, ['--folder', my_folder_path], input=data)
+    arguments = ['existing-mnemonic', '--folder', my_folder_path, '--mnemonic-password', 'TREZOR']
```
Comment (hwwhww) [inline tests/test_cli.py:48]: Should add another test case for `new-mnemonic`.

File: tests/test_key_handling/test_key_derivation/test_mnemonic.py [hunk 2]
```diff
@@ -37,21 +38,27 @@ def test_bip39(language: str, test: Sequence[str]) -> None:
...
+    'language, index, valid',
     [
-        (0, True),
-        (2047, True),
-        (2048, False),
+        ('english', 0, True),
+        ('english', 2047, True),
+        ('english', 2048, False),
     ]
 )
```
Comment (hwwhww) [inline tests/test_key_handling/test_key_derivation/test_mnemonic.py:57]: nitpick: can be in two layers
```python
@pytest.mark.parametrize(
    'language', ['english']
)
@pytest.mark.parametrize(
    'index, valid',
    [
        (0, True),
        (2047, True),
        (2048, False),
    ]
)
```

File: eth2deposit/credentials.py [hunk 3]
```diff
@@ -41,8 +41,10 @@ def __init__(self, *, mnemonic: str, index: int, amount: int, fork_version: byte
         self.signing_key_path = f'{withdrawal_key_path}/0'
 
         # Do NOT use password for seed generation.
```
Comment (hwwhww) [inline eth2deposit/credentials.py:43]: The comment isn't true now.
```suggestion
```

File: eth2deposit/cli/new_mnemonic.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        click.clear()
+        click.echo('This is your seed phrase. Write it down and store it safely, it is the ONLY way to retrieve your deposit.')  # noqa: E501
+        click.echo('\n\n%s\n\n' % mnemonic)
+        click.pause('Press any key when you have written down your mnemonic.')
+
+        click.clear()
+        test_mnemonic = click.prompt('Please type your mnemonic (separated by spaces) to confirm you have written it down\n\n')  # noqa: E501
+        test_mnemonic = test_mnemonic.lower()
+    click.clear()
+    ctx.obj = {'mnemonic': mnemonic, 'mnemonic_password': ''}
```
Comment (hwwhww) [inline eth2deposit/cli/new_mnemonic.py:42]:
```suggestion
    # Do NOT use mnemonic_password.
    ctx.obj = {'mnemonic': mnemonic, 'mnemonic_password': ''}
```

**Review by hwwhww — COMMENTED** [review]
My first pass. 👀  Good job!

**Comment by fabdarice** [comment]
I was not able to re-generate the same public keys using the seed.
To be fair, I remember generating those prior to medalla launch so it's been a while ago..

**Comment by hwwhww** [comment]
@fabdarice 

The BLS library version was IETF draft 02 when Medalla was launched, but now it's **IETF BLS draft 04**. Unfortunately, the KeyGen algorithm was changed when 02->04 so you're unable to use this PR to re-generate the early Medalla keys. 😢 

The IETF BLS draft 04 will be the BLS version for phase 0 launch. You can also use it to re-generate the Zinken testnet keys.

Sorry for the confusion!

**Comment by fabdarice** [comment]
> @fabdarice
> 
> The BLS library version was IETF draft 02 when Medalla was launched, but now it's **IETF BLS draft 04**. Unfortunately, the KeyGen algorithm was changed when 02->04 so you're unable to use this PR to re-generate the early Medalla keys. 
> 
> The IETF BLS draft 04 will be the BLS version for phase 0 launch. You can also use it to re-generate the Zinken testnet keys.
> 
> Sorry for the confusion!

Yep, saw that. No problem. This PR will come in handy for those who lost their validator keys.

**Review by hwwhww — CHANGES_REQUESTED** [review]
- Some missing code during resolving commit conflicts.
- #139 fixes the Dockerfile and Makefile

File: README.md [hunk 5]
```diff
@@ -183,23 +222,36 @@ pip3 install -r requirements.txt
...
+
 ###### Arguments
-See [here](#arguments)
 
-###### Successful message
-See [here](#successful-message)
+See [here](#new-mnemonic-arguments) for `new-mnemonic` arguments
+See [here](#existing-mnemonic-arguments) for `existing-mnemonic` arguments
 
 #### Option 4. Use Docker image
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply README.md:256]: I pushed #139 :)

[the remark this answers is not in the document]
Comment (CarlBeek) [reply README.md:261]: Merged, thanks @hwwhww!

File: eth2deposit/cli/generate_keys.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    '''
+    This is a decorator that, when applied to a parent-command, implements the
+    to obtain the necessary arguments for the generate_keys() subcommand.
+    '''
+    decorators = [
+        click.option(
+            '--num_validators',
+            prompt='Please choose how many validators you wish to run',
+            required=True,
+            type=click.IntRange(0, 2**32),
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:35]: You are correct, both are inclusive bounds. Weird that it doesn't follow the python standards eg`x[0:2**32]`

File: eth2deposit/cli/generate_keys.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            type=click.Path(exists=True, file_okay=False, dir_okay=True),
+            default=os.getcwd()
+        ),
+        click.option(
+            '--chain',
+            prompt='Please choose the (mainnet or testnet) network/chain name',
+            type=click.Choice(ALL_CHAINS.keys(), case_sensitive=False),
+            default=MAINNET,
+        ),
+        click.password_option('--keystore_password', prompt='Type the password that secures your validator keystore(s)')
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:48]: Apologies, I had something really weird going on with my git setup. I'll make sure it gets in there.

[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py]: Addressed with https://github.com/ethereum/eth2.0-deposit-cli/pull/123/commits/29f00891fe11cf1027f3bcecdaeca33454d3e507

File: README.md [hunk 8]
```diff
@@ -71,21 +79,30 @@ See [releases page](https://github.com/ethereum/eth2.0-deposit-cli/releases) to
...
 ```
 
-###### Arguments
+###### Commands
+
+The CLI offers different commands depending on what you want to do with the tool.
+
+| Command | Description |
+| ------- | ----------- |
+| `new-mnemonic` | (Recommended) If you don't already have a mnemonic that you have securely backed up, or you want to have a separate mnemonic for your eth2 validators, use this option. |
```
Comment (hwwhww) [inline README.md:100]: simplification suggestion:

```suggestion
| `new-mnemonic` | (Recommended) This command is used to generate keystores with a new mnemonic. |
```

File: README.md [hunk 9]
```diff
@@ -71,21 +79,30 @@ See [releases page](https://github.com/ethereum/eth2.0-deposit-cli/releases) to
...
+| `existing-mnemonic` | If you have a mnemonic that you already use, then this option allows you to derive new keys from your existing mnemonic. Use this tool, if you have already generated keys with this CLI before, if you want to reuse your mnemonic that you know is secure that you generated elsewhere (reusing your eth1 mnemonic etc), or if you lost your keystores and need to recover your validator/withdrawal keys. |
```
Comment (hwwhww) [inline README.md:101]: simplification suggestion:


```suggestion
| `existing-mnemonic` | This command is used to re-generate or derive new keys from your existing mnemonic. Use this command, if (i) you have already generated keys with this CLI before, (ii) you want to reuse your mnemonic that you know is secure that you generated elsewhere (reusing your eth1 mnemonic .etc), or (iii) you lost your keystores and need to recover your keys. |
```

File: README.md [hunk 10]
```diff
@@ -94,11 +111,22 @@ You can use `--help` flag to see all arguments.
...
+###### `existing-mnemonic` Arguments
+
+You can use `existing-mnemonic --help` to see all arguments. Note that if there are missing arguments that the CLI needs, it will ask you for them.
+
+| Argument | Type | Description |
+| -------- | -------- | -------- |
+| `--validator_start_index` | Non-negative integer | The index of the first validator's keys you wish to generate. If this is your first time generating keys with this mnemonic, use 0. If you have generated keys using this mnemonic before, use the next index from which you want to start generating keys from (eg, if you've generated 4 keys before (keys #0, #1, #2, #3), then enter 4 here.|
+| `--num_validators`  | Non-negative integer | The number of signing keys you want to generate. Note that the child key(s) are generated via the same master key. |
+| `--folder` | String. Pointing to `./validator_keys` by default | The folder path for the keystore(s) and deposit(s) |
+| `--chain` | String. `mainnet` by default | The chain setting for the signing domain. |
```
Comment (hwwhww) [inline README.md:123]: It looks good to add the descriptions of other arguments to`"@click.option(..., help="<description>")` as well (currently, only the `--mnemonic-password` has it)

Comment (CarlBeek) [reply README.md:123]: https://github.com/ethereum/eth2.0-deposit-cli/pull/123/commits/d3149c688a30f4af33ff0344730757dbda944f1a Added help parameters for all the commands and arguments.

File: README.md (hunk 9 above)
Comment (CarlBeek) [reply README.md]: resolved via https://github.com/ethereum/eth2.0-deposit-cli/pull/123/commits/10bc333ee973d6243e9bfb7fb179b29f373ce7e8

File: README.md (hunk 8 above)
Comment (CarlBeek) [reply README.md]: resolved via https://github.com/ethereum/eth2.0-deposit-cli/pull/123/commits/10bc333ee973d6243e9bfb7fb179b29f373ce7e8

File: eth2deposit/cli/existing_mnemonic.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    type=str,
+)
+@click.password_option(
+    '--mnemonic-password',
+    default='',
+    help=('This is almost certainly not the argument you are looking for: it is for mnemonic passwords, not keystore '
+          'passwords. Providing a password here when you didn\'t use one initially, can result in lost keys (and '
+          'therefore funds)! Also note that if you used this tool to generate your mnemonic intially, then you did not '
+          'use a mnemonic password. However, if you are certain you used a password to "increase" the security of your '
+          'mnemonic, this is where you enter it.mnemonic'),
```
Comment (hwwhww) [inline eth2deposit/cli/existing_mnemonic.py:46]:
```suggestion
          'mnemonic, this is where you enter it.'),
```

**Review by hwwhww — APPROVED** [review]
I pushed some minor fixes her and it LGTM 👍 

#143 added the cli key regeneration test case.

**Comment by antliv** [comment]
U burnt me

**Pull request merged**